op_crates/crypto/01_crypto.js

```diff
@@ -1,25 +1,211 @@
+    if (!ArrayBuffer.isView(arrayBufferView)) {
+      throw new TypeError(
+        "Argument 1 does not implement interface ArrayBufferView",
+      );
+    }
+    if (
+      !(
+        arrayBufferView instanceof Int8Array ||
+        arrayBufferView instanceof Uint8Array ||
+        arrayBufferView instanceof Int16Array ||
+        arrayBufferView instanceof Uint16Array ||
+        arrayBufferView instanceof Int32Array ||
+        arrayBufferView instanceof Uint32Array ||
+        arrayBufferView instanceof Uint8ClampedArray
+      )
+    ) {
+      throw new DOMException(
+        "The provided ArrayBufferView is not an integer array type",
+        "TypeMismatchError",
+      );
+    }
+    if (arrayBufferView.byteLength > 65536) {
+      throw new DOMException(
+        `The ArrayBufferView's byte length (${arrayBufferView.byteLength}) exceeds the number of bytes of entropy available via this API (65536)`,
+        "QuotaExceededError",
+      );
     }
     const ui8 = new Uint8Array(
-      typedArray.buffer,
-      typedArray.byteOffset,
-      typedArray.byteLength,
+      arrayBufferView.buffer,
+      arrayBufferView.byteOffset,
+      arrayBufferView.byteLength,
     );
     core.jsonOpSync("op_get_random_values", {}, ui8);
-    return typedArray;
+    return arrayBufferView;
   }
+
+  // Algorithm normalization, which involves storing the expected data for
+  // each pair of algorithm and crypto operation, is done on the JS side because
+  // it is convenient.
+  // Shall it stay that way, or are we better to move it on the Rust side?
+
+  // We need this method after initialization, anytime we need to normalize
+  // a provided algorithm. We store it here to prevent prototype pollution.
+  const toUpperCase = String.prototype.toUpperCase;
+
+  class RegisteredAlgorithmsContainer {
+    #nameIndex;
+    #definitions;
+
+    constructor(definitions) {
+      this.#nameIndex = Object.create(null);
+      this.#definitions = Object.create(null);
+      for (const [name, definition] of Object.entries(definitions)) {
+        this.#nameIndex[name.toUpperCase()] = name;
+        this.#definitions[name] = definition;
+      }
+    }
+
+    getDefinition(name) {
+      const upperCaseName = toUpperCase.call(name);
+      const registeredName = this.#nameIndex[upperCaseName];
+      if (registeredName === undefined) {
+        return undefined;
+      }
+      return this.#definitions[registeredName];
+    }
+
+    normalizeName(providedName) {
+      const upperCaseName = toUpperCase.call(providedName);
+      return this.#nameIndex[upperCaseName];
+    }
+  }
+
+  const supportedAlgorithms = {
+    digest: new RegisteredAlgorithmsContainer({
+      "SHA-1": {},
+      "SHA-256": {},
+      "SHA-384": {},
+      "SHA-512": {},
+    }),
+  };
+
+  const digestByteLengths = {
+    "SHA-1": 20,
+    "SHA-256": 32,
+    "SHA-384": 48,
+    "SHA-512": 64,
+  };
+
+  function normalizeAlgorithm(algorithm, registeredAlgorithms) {
```

Comment:
Doing all that work in JS because:
- It's easier
- We only ever have to serialise algorithm parameters that are normalised, meaning we will never attempt to serialise the input as provided by the user (which could be huge, be of a weird type, or contain circular references)